core/include/drivers/rstctrl.h

```diff
@@ -0,0 +1,217 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2021, Linaro Limited
+ *
+ * Self contained reset controller framework in rstctrl.h
+ * A reset controller exposes reset level controls to a
+ * platform.
+ */
+
+#ifndef __DRIVERS_RSTCTRL_H
+#define __DRIVERS_RSTCTRL_H
+
+#include <kernel/dt.h>
+#include <kernel/dt_driver.h>
+#include <stdint.h>
+#include <tee_api_types.h>
+
+struct rstctrl;
+
+struct rstctrl_ops {
+	/*
+	 * Operators on reset control(s) exposed by a reset controller
+	 *
+	 * assert_level: Assert reset level on control with a timeout hint
+	 * deassert_level: Deassert reset level on control with a timeout hint
+	 * get_name: Get a string name for the controller, or NULL is none
+	 *
+	 * @rstctrl: Reset control reference
+	 * @id: Identifier for the reset level control in the reset controller
+	 * @to_ms: Timeout in microsencond or RSTCTRL_NO_TIMEOUT, may be ignored
+	 * by reset controller.
+	 * Return a TEE_Result compliant code
+	 */
+	TEE_Result (*assert_level)(struct rstctrl *rstctrl, unsigned int to_us);
+	TEE_Result (*deassert_level)(struct rstctrl *rstctrl,
+				     unsigned int to_us);
+	const char *(*get_name)(struct rstctrl *rstctrl);
+};
+
+/*
+ * struct rstctrl - Instance of a control exposed by a reset controller
+ * @ops: Operators of the reset controller
+ * @priv: Reset control private data (unsigned value or memory pointer)
+ * @exclusive: Set when a consumer has exclusive control on the reset level
+ */
+struct rstctrl {
+	const struct rstctrl_ops *ops;
+	uintptr_t priv;
```

Comment:
Do we need this field? I guess it wouldn't if we intended to include this struct in a larger device specific struct.

Reply:
Indeed, could be removed.

Reply:
do we restore this generic private ref for the driver?

In current series state, it is up to the platform reset controller driver to bind the struct address to its local reset line reference. 

Example: planned for [_stm32_reset.c_ (ephemeral link)](https://github.com/etienne-lms/optee_os/commit/2e12ef74f202562574fff9dc42009d3145b73d67?diff=unified#diff-1ceec50dd256ff8eba2db4dd1c4f52dd1ba4214b40b3e5af0ea89966c4748d1bR22)
```c
/* Exposed rstctrl instance */
struct stm32_rstline {
	unsigned int id;
	struct rstctrl rstctrl;
	SLIST_ENTRY(stm32_rstline) link;
};

static struct stm32_rstline *to_rstline(struct rstctrl *rstctrl)
{
	assert(rstctrl);

	return container_of(rstctrl, struct stm32_rstline, rstctrl);
}
```

Adding `priv` in `struct rstctrl` increases the struct size, may be be easier for platforms willing to save a reference (local pointer or numeric identifer.
Note the numeric identifer must be set by the provider driver: it may be the DT binding Id for 1 cell `#reset-cells` providers, it may need a conversion when DT binding rely or more IDs (`#reset-cells = 2;`).